pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -403,8 +463,32 @@ func adsPushAll() {
 	// It will include sending all configs that envoy is listening for, including EDS.
 	// TODO: get service, serviceinstances, configs once, to avoid repeated redundant calls.
 	// TODO: indicate the specific events, to only push what changed.
-	for _, client := range tmpMap {
-		client.pushChannel <- &XdsEvent{}
+	for _, c := range tmpMap {
+		// Using non-blocking push has problems if 2 pushes happen too close to each other
+		client := c
+		//go func(client *XdsConnection) {
+		to := time.After(PushTimeout)
+		select {
+		case client.pushChannel <- &XdsEvent{}:
+			client.LastPush = time.Now()
+			client.LastPushFailure = timeZero
+		case <-client.doneChannel: // connection was closed
+		case <-to:
+			pushTimeouts.Add(1)
+			//default:
+			// This may happen to some clients if the other side is in a bad state and can't receive.
+			// The tests were catching this - one of the client was not reading.
+			if client.LastPushFailure.IsZero() {
+				client.LastPushFailure = time.Now()
+				log.Infof("Failed to push, client busy %s", client.ConID)
+			} else {
+				if time.Since(client.LastPushFailure) > 10*time.Second {
+					log.Infof("Repeated failure to push %s", client.ConID)
```

Comment:
Should this be a warning?

Reply:
Right. And forgot to add a counter. 

Please request metrics on any interesting error path !